File: 0238-product-of-array-except-self/0238-product-of-array-except-self.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int n = nums.length;
        int[] answer = new int[n];

        int leftproduct = 1;
        int rightproduct = 1;

        for (int i = 0; i < n; i++) {
            answer[i] = leftproduct;
            leftproduct *= nums[i];
        }

        for (int i = n - 1; i >= 0; i--) {
            answer[i] *= rightproduct;
            rightproduct *= nums[i];
        }

        return answer;
    }
}

File: 0238-product-of-array-except-self/0238-product-of-array-except-self.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        vector<int> leftproducts(nums.size());
        vector<int> rightproducts(nums.size());

        vector<int> finalarr(nums.size());

        int leftproduct = 1;
        int rightproduct = 1;

        for(int i = 0; i < nums.size(); i++) {
            leftproducts.at(i) = leftproduct;
            leftproduct *= nums.at(i);
        }

        for(int i = nums.size() - 1; i >= 0; i--) {
            rightproducts.at(i) = rightproduct;
            rightproduct *= nums.at(i);
        }

        for(int i = 0; i < nums.size(); i++) {
            finalarr.at(i) = leftproducts.at(i) * rightproducts.at(i);
        }

        return finalarr;
    }
};
